### PR TITLE
✨ Source Facebook-Marketing: Add Selectable Auth

### PR DESCRIPTION
**Pull request by JohnMikkelsonOvative**
## What
The Source Facebook Marketing Connector does not allow for choosing between providing an access token or using oauth, it only allows for users to define a source connection by following an oauth flow. This does not allow for use-cases in which a large organization may provision an access token to be used to authenticate to the Facebook Marketing data, rather than giving a user's personal Facebook account access.

## How
The `spec` defines a `credentials` field that is a union between newly created `OAuthCredentials` and `ServiceAccountCredentials`. This allows for the user to select between using the Oauth flow and inserting Service Account Credentials.

## Review guide
1. spec.py
2. config.py
3. source.py
4. sample_config.json
5. metadata.yaml
6. pyproject.toml
8. request_builder.py
9. test_ads_insights_action_product_id.py
10. test_videos.py
11. test_source.py

## User Impact
The user is able to select between using the Oauth flow and inserting Service Account Credentials when setting up the Facebook Marketing Source.

## Can this PR be safely reverted and rolled back?

- [ ] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by JohnMikkelsonOvative** [comment]
During the migration to `credentials` selectable configuration, the decision was made to change the naming of `ACCESS_TOKEN` to `SERVICE_ACCOUNT_INFO`. This was due to the same naming convention being used in the [Source-Google-Drive](https://github.com/JohnMikkelsonOvative/airbyte/blob/master/airbyte-integrations/connectors/source-google-drive/source_google_drive/spec.py). Does this make sense to the Airbyte team, or should we change that naming convention to something more agnostic?

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 1]
```diff
@@ -95,7 +95,10 @@ def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) ->
             if config.start_date and config.end_date < config.start_date:
                 return False, "End date must be equal or after start date."
 
-            api = API(access_token=config.access_token, page_size=config.page_size)
+            if config.credentials.auth_type == 'Client':
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:98]: Just small suggestion is to use maybe Enum class to represent the auth type value.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Thanks for the contribution @JohnMikkelsonOvative. It was added to the Connector team backlog and it will be reviewed in future sprint. I'll check with the project manager a timeline to make aware, ask your patience until the moment of the review. As this impact in a critical connector the review can take some time and have some back and forward, the goal is to introduce new features with the lower impact possible. In any time, feel free to reach me out in Slack or ping here to get update about the review process.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py [hunk 2]
```diff
@@ -95,7 +95,10 @@ def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) ->
...
+                api = API(access_token=config.credentials.refresh_token, page_size=config.page_size)
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/source.py:99]: The OAuth flow, will return the `access_token` only. 
Based on this [line of code](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/flows/facebook/FacebookOAuthFlow.java#L31)

File: airbyte-integrations/connectors/source-facebook-marketing/unit_tests/integration/config.py [hunk 3]
```diff
@@ -25,7 +25,10 @@ class ConfigBuilder:
     def __init__(self) -> None:
         self._config: MutableMapping[str, Any] = {
             "account_ids": [ACCOUNT_ID],
-            "access_token": ACCESS_TOKEN,
+            "credentials": {
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-facebook-marketing/unit_tests/integration/config.py:28]: There is no need to replace the tests, but `extend` them with the new functionality, using parametrized test case. With this changes, we loose the backward compatibility tests for the existing Customers.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py [hunk 4]
```diff
@@ -27,6 +28,44 @@
...
+        title="Client ID",
+        description="Client ID for the Facebook Marketing API",
+        airbyte_secret=True,
+    )
+    client_secret: str = Field(
+        title="Client Secret",
+        description="Client Secret for the Facebook Marketing API",
+        airbyte_secret=True,
+    )
+    refresh_token: str = Field(
```
Comment (bazarnov) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:47]: The OAuth2.0 Flow doesn't return the `refresh_token`, could you please refer to the official doc, where this is possible? Thank you.

**Review by bazarnov — CHANGES_REQUESTED** [review]
Requesting changes, because this change will break the existing Customers.

The path should look like this:
1) Support both `old` and `new` way of authentication to cover the existing Customers
2) Write the test cases for both scenarios
3) Make it possible to migrate existing customers to use the `new` way of authentication by applying the `config migration`, see the example in the `/config_migrations.py`
4) Release this update (at this point, all existing Customers are covered with the config_migration, this will allow to win the time, before we update the OAuth Flow for the platform side, [here](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/flows/facebook/FacebookOAuthFlow.java#L128-L131) by removing this method.
5) Done; the platform will return the `access_token` back to the `credentials.access_token` as expected after point `4` is released, which means we are good to go with the new authentication way of things.

**Comment by maxi297** [comment]
@bazarnov For my personal knowledge, what does removing [getDefaultOAuthOutputPath](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/BaseOAuthFlow.java#L126-L133) do? It is also [deprecated in the interface](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/BaseOAuthFlow.java#L126-L133) which seems like we should move away from that

**Comment by bazarnov** [comment]
> @bazarnov For my personal knowledge, what does removing [getDefaultOAuthOutputPath](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/BaseOAuthFlow.java#L126-L133) do? It is also [deprecated in the interface](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/BaseOAuthFlow.java#L126-L133) which seems like we should move away from that

You've looked to the wrong place, see [this](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/BaseOAuth2Flow.java#L372-L376)
Which is inherited from [this](https://github.com/airbytehq/airbyte-platform-internal/blob/6ec42b2340bce9f690fd3518c38fd06a2ad5762b/oss/airbyte-oauth/src/main/java/io/airbyte/oauth/flows/facebook/FacebookOAuthFlow.java#L27)

**Comment by maxi297** [comment]
Note that this method is in `FacebookOAuthFlow` and this will affect Instagram and Facebook Pages as well. Is this an issue? We can probably re-implement this in the specific classes for those connectors

**Comment by bazarnov** [comment]
> Note that this method is in `FacebookOAuthFlow` and this will affect Instagram and Facebook Pages as well. Is this an issue? We can probably re-implement this in the specific classes for those connectors

We can add the existing method from the base class to all of the other Facebook-related OAuth flows to keep them running `as is` for now, then create the follow-up issues, to modify each of them. In this way, the Instagram and Facebook Pages, will not be affected with the platform change.

**Comment by bazarnov** [comment]
@cmm-airbyte Could you please fix the conflicts and ensure we pass the CAT successfully before this is finally approved? Thanks

**Pull request closed without merging**